### PR TITLE
Set3

### PR DESCRIPTION
**Pull request by MarGabor**
Pull request to merge sEt3 into master.
-python code to automatically read IDs from spreadsheet and get names and links of publications from author page in webofscience,com via javascript execution
-manual log in required
-error logging
-data generated added to repo

### CODE REVIEW

**Comment by MarGabor** [comment]
Habe ausversehen auch "lazlostrange" als reviewer angefragt. Kann ignoriert werden.

**Comment by MarioStanke** [comment]
Why are there only 31 publication lists? Did the others fail?

**Comment by MarGabor** [comment]
> Why are there only 31 publication lists? Did the others fail?

It must be that the others have either no publications listed on WOS or have no WOS-ID at all.

**Comment by MarGabor** [comment]
I'll have a look at the dataAcquisitionLog.txt and see if I can find a page, where it timed out although there are publications listed. But I personally suspect that people have not actually created an accoount on WOS.

**Comment by MarioStanke** [comment]
For example Roland Pulch is missing and has publications:
https://www.webofscience.com/wos/author/record/29340731

**Comment by MarioStanke** [comment]
We should have 86 publication lists because WoS has only authors (by definition >0 publications).

**Comment by MarGabor** [comment]
Mhh strange. I supervised the process and regularly there was a page that said that "this page is algorithmically generated". Maybe some kind of covert bot protection?

**Comment by MarGabor** [comment]
Actually I can't see the page of Roland Pulch, either.

**Comment by MarGabor** [comment]
Will push new changes. No merging required yet. Managed to get 102 lists extracted this time with use of VPN and institutional login. The problem with that seems to be that I get a random "Access denied" error most of the time when I log in in the headless browser. So this is far from being consistent. But I managed to log in at least once and could extract almost all of the data. (s. dataAcquisitionLog.txt for details, once changes are pushed). I'm trying to push the last two remaining literature lists, too. But my access is denied right now, so I will try again later. Will request new review once I think it's finished.

**Comment by MarioStanke** [comment]
Good. Please let me explicitly know when I should review (Will push new changes. No merging required yet.)

**Comment by MarGabor** [comment]
Locally merged. Review can begin.

File: src/dataAcquisition/dataAcquisition.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+    with open(logPath,'a') as file:
+            file.write("["+dtstr+"]\tFiltered to "
+                       +str(len(newLinklist))+" links.\n")
+    return newLinklist
+
+#main function for testing
+def main():
+    #js code to be converted
+    linkgrabber = getJsCode("js_scripts/linkgrabber.js")
+    spreadsheet = pandas.read_csv("../../data/ProfsGW-proflist.csv")
```
Comment (MarioStanke) [inline src/dataAcquisition/dataAcquisition.py:110]: Should be `prof.tbl`.

File: src/dataAcquisition/dataAcquisition.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+    sub_df = getNamesAndURLs(spreadsheet)
+    #geckodriver location can differ
+    driver = launchFirefox('/snap/bin/firefox.geckodriver')
+    """
+    Use time to:
+    1. Make sure all sessions for WOS are ended in other browsers.
```
Comment (MarioStanke) [inline src/dataAcquisition/dataAcquisition.py:116]: Copy or move to README.md

File: src/dataAcquisition/dataAcquisition.py [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+            file.write("["+dtstr+"]\tError while executing script. "
+                       +type(Exception).__name_+"\n")
+    finally:
+        with open(logPath,'a') as file:
+            file.write("["+dtstr+"]\tExtracted "+str(len(result))+" links.\n")
+        return result
+
+#extract names and ID from spreadsheet and create data frame with URLs
+def getNamesAndURLs(df: pandas.DataFrame) -> pandas.DataFrame:
+    names_IDs_URLs = df[["Vorname","Nachname","ID"]]
```
Comment (MarioStanke) [inline src/dataAcquisition/dataAcquisition.py:90]:
```suggestion
    names_IDs_URLs = df[["firstname","lastname","WoSID","id"]]
```
id is ours, WoSID is from WebOfScience

File: src/dataAcquisition/dataAcquisition.py [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+    4. Log in via your institution to gain full access to website
+       make also sure to tick the box that says "Remember me for this session"
+       When "Access denied" error occurs, delete cookies manually and resend
+       the form (i.e. refresh page) afterwards.
+    5. Accept cookies and dismiss other pop-ups
+    6. Wait until program is finished and you regain control over shell
+    """
+    time.sleep(120)
+    #i is index of the spreadsheet
+    i = 1
```
Comment (MarioStanke) [inline src/dataAcquisition/dataAcquisition.py:129]: not `i` but `sub_df["id"]`

File: src/dataAcquisition/dataAcquisition.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+                       +sub_df.loc[sub_df.ID == ID, "Nachname"].values[0]
+                       +","+str(int(ID))+";\n\n")
+            #write links
+            for link in linklist:
+                file.write(link[0]+"\t"+link[1]+"\n\n")
+    driver.quit()
+    return 0
+
+if __name__ == "__main__":
+    main()
```
Comment (MarioStanke) [inline src/dataAcquisition/dataAcquisition.py:156]: newline

File: src/dataAcquisition/js_scripts/linkgrabber.js [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+var x = document.querySelectorAll("a");
+var myarray = []
+for (var i=0; i<x.length; i++){
+    var nametext = x[i].textContent;
+    var cleantext = nametext.replace(/\s+/g, ' ').trim();
+    var cleanlink = x[i].href;
+    myarray.push([cleantext,cleanlink]);
+};
+return myarray;
+//}
```
Comment (MarioStanke) [inline src/dataAcquisition/js_scripts/linkgrabber.js:13]: newline

File: src/dataAcquisition/dataAcquisition.py [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+#visit website specified under URL and wait for specific element
+def navigateToWebsite(driver, URL: str,
+                      elementToWaitFor: str) -> webdriver.Firefox:    
+    dt = datetime.datetime.now()
+    dtstr = dt.strftime('%d/%m/%Y %H:%M:%S')
+    try:
+        driver.get(URL)
+    except:
+        with open(logPath,'a') as file:
+            file.write("["+dtstr+"]\tError: Page \""+URL
```
Comment (MarioStanke) [inline src/dataAcquisition/dataAcquisition.py:49]: For spacing see coding style.

File: src/dataAcquisition/js_scripts/wos_gotItClicker.js [hunk 8]
```diff
@@ -0,0 +1,5 @@
+//trigger gotIt event
+//function  gotItClicker(){
+    gotItButton = document.getElementById("loadingButton");
+    gotItButton.click();
+//}
```
Comment (MarioStanke) [inline src/dataAcquisition/js_scripts/wos_gotItClicker.js:5]: \n

**Review by MarioStanke — CHANGES_REQUESTED** [review]
Good job. 
However, please use the IDs from `data/prof.tbl`, e.g. Christiane Helm has ID 8.
Please put a description, how it is run in `src/Readme.md`.

File: src/dataAcquisition/dataAcquisition.py (hunk 7 above)
Comment (MarioStanke) [reply src/dataAcquisition/dataAcquisition.py:49]: Coding style. Spaces around +
```
# Correct:
i = i + 1
submitted += 1
x = x*2 - 1
hypot2 = x*x + y*y
c = (a+b) * (a-b)
# Wrong:
i=i+1
submitted +=1
x = x * 2 - 1
hypot2 = x * x + y * y
c = (a + b) * (a - b)
```

File: src/dataAcquisition/dataAcquisition.py [hunk 9]
```diff
@@ -47,24 +47,26 @@ def navigateToWebsite(driver, URL: str,
     except:
         with open(logPath,'a') as file:
             file.write("["+dtstr+"]\tError: Page \""+URL
```
Comment (MarioStanke) [inline src/dataAcquisition/dataAcquisition.py:49]:
```suggestion
            file.write("[" + dtstr + "]\tError: Page \"" + URL
```

**Review by MarioStanke — APPROVED** [review]
Looks good. There was just one unanswered comment. I will correct the coding style there myself.

**Pull request merged**